### PR TITLE
Add (KD)TreeIndex

### LINKED ISSUE #475: API design for pointwise indexing
There have been a number of threads discussing possible improvements/extensions to `xray` indexing. The current indexing behavior for `isel` is orthogonal indexing - in other words, each coordinate is treated independently (see #214 and #411 for more discussion).

So the question: what is the best way to incorporate diagonal or pointwise indexing in `xray`? I see two main goals / applications:
1. support simple form of [_`numpy`_ style integer array indexing](http://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#purely-integer-array-indexing)
2. support pointwise array indexing along coordinates via computation of nearest-neighbor indexes - I think this can also be thought of as a form of resampling.

Input from @WeatherGod, @wholmgren, and @shoyer would be great.

### LINKED ISSUE #10157: Selecting point closest to (lon0, lat0) when lon,lat coordinates are 2D
### Is your feature request related to a problem?

It's very common to want to extract a time series at a specified coordinate location, and I'm wondering whether xarray could support this directly without using xoak.   Currently I'm using xoak, as in this reproducible example:
``` python
import xarray as xr
import intake
import xoak

intake_catalog_url = 'https://usgs-coawst.s3.amazonaws.com/useast-archive/coawst_intake.yml'
cat = intake.open_catalog(intake_catalog_url)
ds = cat['COAWST-USEAST'].to_dask()

lat, lon = 42.5, -70.0  # Gulf of Maine, 100km east of Boston, MA

da = ds['Hwave']
da.xoak.set_index(['lat_rho', 'lat_rho'], 'scipy_kdtree')
ds_point = xr.Dataset({"lon": ("point", [lon]), "lat": ("point", [lat])})
da.xoak.sel(lat_rho=ds_point.lat, lon_rho=ds_point.lon).sel(ocean_time='2012-10-01')
```
Would it be possible to enable this in xarray or is this too specific a functionality to consider?

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by benbovy**
- [x] Closes #10157, closes #475
- [ ] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`
- [ ] Add basic documentation
- [ ] `.sel()` with scalar values

This PR adds an `xarray.indexes.TreeIndex` for point-wise label-based  indexing of irregular data (n-dimensional coordinates). By default it is using `scipy.spatial.KDTree` under the hood.

It is heavily inspired from [xoak](https://github.com/xarray-contrib/xoak/). Actually it could eventually replace it, or alternatively xoak could be built on top of this. There's no support for (dask) chunked coordinates here, although in xoak that feature is experimental and not working well anyway.

`TreeIndex` shares some similarities with `CoordinateTransformIndex` in its implementation and supported features:

- Like for `CoordinateTransform`, there's a small `TreeAdapter` abstraction layer that allows plugging in alternative index objects like `sklearn.neighbors.KDTree`, `sklearn.neighbors.BallTree`, etc. We can keep this private for now and expose it later.
- There's some rudimentary support for (exact) alignment. Comparing `scipy.spatial.KDTree` objects is done here based on the fact that they hold a view to their original data points (i.e., that's basically equivalent to comparing the associated coordinate variables).
- Only advanced label-based selection (i.e., point-wise selection using `DataArray` or `Variable` objects) is supported.

### CODE REVIEW

**Comment by dcherian** [comment]
Nice! Would be good to add a section to this [gallery example](https://docs.xarray.dev/en/stable/examples/ROMS_ocean_model.html)

**Comment by benbovy** [comment]
> Would be good to add a section to this [gallery example](https://docs.xarray.dev/en/stable/examples/ROMS_ocean_model.html)

While I agree that it would be nice to showcase `TreeIndex` in the example gallery, I'm not sure that the specific ROMS ocean model example (based on lat/lon 2D coordinates) will be relevant with a (default) KDTree and no longitude wraparound. It will make more sense to use something like xoak's [sklearn_geo_balltree](https://xoak.readthedocs.io/en/latest/_api_generated/xoak.index.sklearn_adapters.SklearnGeoBallTreeAdapter.html) or [s2point](https://xoak.readthedocs.io/en/latest/_api_generated/xoak.index.s2_adapters.S2PointIndexAdapter.html).

**Comment by benbovy** [comment]
This is ready for review.

I can do one pass over the doc's user guide and examples to see if we can use or mention this. I can also do that in a follow-up PR.

Any better idea for the index name than `TreeIndex`? I'm not a big fan of it. `NDPointIndex`? `KDPointIndex`?

**Comment by benbovy** [comment]
We can also see later if there's a need to implement `TreeIndex.concat`, `TreeIndex.stack`, `TreeIndex.isel`... and if yes how best to do so. One concern I have with automatically re-building and propagating a TreeIndex is that building such index may be expensive (likely more than other kinds of indexes).

**Comment by benbovy** [comment]
> Any better idea for the index name than TreeIndex? I'm not a big fan of it. NDPointIndex? KDPointIndex?

I renamed it to `NDPointIndex` that I like better.

The index inline repr now displays the underlying structure:

```
    <xarray.Dataset> Size: 64B
    Dimensions:  (y: 2, x: 2)
    Coordinates:
      * xx       (y, x) float64 32B 1.0 2.0 3.0 0.0
      * yy       (y, x) float64 32B 11.0 21.0 29.0 9.0
    Dimensions without coordinates: y, x
    Data variables:
        *empty*
    Indexes:
      ┌ xx       NDPointIndex (ScipyKDTreeAdapter)
      └ yy
```

**Comment by dcherian** [comment]
Very nice! At the meeting, there was uniform support for merging this.

**Comment by benbovy** [comment]
A few ideas for further improvements (I'll reference this comment in a new issue):

- Alignment with tolerance
  - This is a good case
  - Until Xarray's API allows providing a tolerance value when aligning objects, we could allow providing it as an option at index creation.
  - Picking a tolerance value among the ones set for each index to align would depend on the join method, e.g., pick the smallest value for `"exact"` and `"inner"` and pick the largest value for `"outer"`.
  - For tolerance > 0, we could rely Scipy KDTree's `count_neighbors`. For tolerance == 0 we could just compare the data points as implemented in this PR.

- Join and re-index
  - I guess it is possible to support it via querying all matching point pairs within the given tolerance.
  - Raise an error if tolerance == 0 ?
  - We could rely Scipy KDTree's `query(..., distance_upper_bound=tolerance)`. `query_ball_tree()` could also be a good candidate but it returns nested Python lists that might be slow to process.

- Positional indexing (isel)
  - Currently (this PR) the index is dropped after `isel`, but we could create a new one (from-scratch) instead. 
  - Implementation should be pretty straightforward, we can transform the input dimension indexers into a flat indexer for the (n_points, n_coordinates) point data array via `numpy.ravel_multi_index`.
  - Perhaps make this opt-in via an option at index creation ?

- Concat
  - I guess numpy provides all the utilities needed to implement it in a clever way so that the concatenated (n_points, n_coordinates) point data array stay consistent with the location of the concat dimension `dim` in the original coordinate variables.

- Stack
  - Would it be a useful alternative to `pandas.MultiIndex` for floating-point coordinates, even though the original distribution of the data points is regular?
  - Implementation is trivial
  - Cannot support `unstack` (it only works with `pandas.MultiIndex`)

- Interpolation
  - This is a good case for n-dimensional nearest neighbor, linear, IDW, etc. interpolation.
  - We'd need to figure out first how to plug custom indexes with Xarray's `interp` API.

- GroupBy
  - It would be great to have specialized grouper objects like `KNearestNeighborsGrouper(points: xr.DataArray, k: int)` and `BallNeigborsGrouper(points: xr.DataArray, radius: float)` (@dcherian).

- `neighbors` accessor?
  - It would be nice to have some Xarray-friendly API to, e.g., return the distances to the nearest neighbors, select k-nearest neighbors, etc.

**Comment by benbovy** [comment]
Integration with `xoak`: https://github.com/xarray-contrib/xoak/pull/44

File: xarray/tests/test_nd_point_index.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        ds.sel(xx=1.1, method="nearest")
+
+    with pytest.raises(ValueError, match="invalid label value"):
+        # invalid array-like dimensions
+        ds.sel(xx=[1.1, 1.9], yy=[3.1, 3.9], method="nearest")
+
+    with pytest.raises(ValueError, match=".*dimensions.*conflict"):
+        ds.sel(
+            xx=xr.Variable("u", [1.1, 1.1, 1.1]),
+            yy=xr.Variable("v", [3.1, 3.1, 3.1]),
```
Comment (dcherian) [inline xarray/tests/test_nd_point_index.py]: This is technically possible by broadcasting out, correct?

**Review by dcherian — APPROVED** [review]
Excited to see this in. 

The tests look good to me. Later we should create some kind of integration test suite for custom indexes that tests all the various flavors of indexing Xarray supports.

File: xarray/tests/test_nd_point_index.py (hunk 1 above)
Comment (benbovy) [reply xarray/tests/test_nd_point_index.py]: Yes indeed. In [f418167](https://github.com/pydata/xarray/pull/10478/commits/f41816781048f77ef6c38b536f5b8e9966176f48) the input labels are simply broadcasted against each other after converting them to DataArray objects.
